compiler/rustc_middle/src/mir/query.rs

```diff
@@ -38,10 +38,82 @@ pub enum UnsafetyViolationDetails {
     AccessToUnionField,
     MutationOfLayoutConstrainedField,
     BorrowOfLayoutConstrainedField,
-    CallToFunctionWith,
+    CallToFunctionWith { func_did: DefId, body_did_def_id: DefId },
 }
 
 impl UnsafetyViolationDetails {
+    fn missing_features(&self, tcx: TyCtxt<'_>) -> Vec<Symbol> {
+        use UnsafetyViolationDetails::*;
+        match self {
+            CallToFunctionWith { func_did, body_did_def_id } => {
+                let callee_features = &tcx.codegen_fn_attrs(func_did).target_features;
+                let self_features = &tcx.body_codegen_attrs(*body_did_def_id).target_features;
+
+                let missing_features = callee_features
+                    .iter()
+                    .filter(|feature| !self_features.contains(feature))
+                    .cloned()
+                    .collect::<Vec<Symbol>>();
+                missing_features
+            }
+            _ => Vec::<Symbol>::new(),
+        }
+    }
+    pub fn note_missing_feature(&self, tcx: TyCtxt<'_>) -> String {
+        use UnsafetyViolationDetails::*;
+        match self {
+            CallToFunctionWith { func_did: _, body_did_def_id: _ } => {
+                let missing_features = self.missing_features(tcx);
+
+                let target_features = &tcx
+                    .sess
+                    .target_features
+                    .iter()
+                    .filter(|feature| missing_features.contains(feature))
+                    .map(|f| f.as_str())
+                    .collect::<Vec<&str>>();
+
+                if target_features.is_empty() {
+                    String::new()
+                } else {
+                    format!(
+                        "the {} target feature(s) being enabled in the build configuration does not remove the requirement to list it in `#[target_feature]`",
+                        target_features.join(" "),
+                    )
+                }
+            }
+            _ => String::new(),
+        }
+    }
+
+    pub fn help_missing_feature(&self, tcx: TyCtxt<'_>) -> String {
+        use UnsafetyViolationDetails::*;
+        match self {
+            CallToFunctionWith { func_did: _, body_did_def_id: _ } => {
+                let missing_features = self.missing_features(tcx);
+                let n = if missing_features.len() > 5 { 5 } else { missing_features.len() };
+
+                let missing_features_str = missing_features[..n]
+                    .iter()
+                    .map(|f| f.as_str())
+                    .collect::<Vec<&str>>()
+                    .join(" ");
+
+                let additional_str = if missing_features.len() > 5 {
```

Comment:
Is it useful to elide the rest of the necessary features? I think in this case it's probably best to just list them all -- I don't expect for there to typically be too many to want to list them all?

Reply:
I've asked for this in the issue linked above, because I've thought it's bad to have lists without limiting.

But it seems there are few cases where we'd run into such situations: While there are functions in the standard library which require 5 target features (say `_mm_mask_expandloadu_epi16`), it seems that this is also the maximum number globally according to [this](https://grep.app/search?q=%5C%23%5C%5Btarget_feature%5C%28enable.%2A%2C.%2A%2C.%2A%2C.%2A%2C&regexp=true&filter[lang][0]=Rust) search. Four target features are still used outside of the rust-lang crates ([in the sha1 crate](https://github.com/RustCrypto/hashes/blob/ca60ca9c287eb4c44e6a6e52e81cc75131cc7e2d/sha1/src/compress/x86.rs#L33)), but one can't find 5 or more, at least not inside a single `#[target_feature(...)]` attribute.

So I suppose that we can remove this eliding.

Reply:
Ok, I get it